ops/model.py

```diff
@@ -1348,6 +1349,73 @@ class InvalidStatusError(ModelError):
     """Raised if trying to set an Application or Unit status to something invalid."""
 
 
+def _validate_action_result_key(key: str) -> None:
+    """Use regular expressions to validate that the given key is in the correct format.
+
+    Correct format is defined as starting with and ending with a lowercase alphanumeric
+    character, and containing nothing but lowercase, alphanumeric characters, hyphens and
+    periods.
+
+    Arguments:
+        key: a string representing the key for an item in an action result.
+
+    Raises:
+        ValueError: if a dict is passes with a key that fails to meet the format requirements.
+    """
+    # Existing charms may well be passing dicts to set_result that contain dicts pre-formatted
+    # to use dotted notation having worked around this in the past. We will deprecate this in the
+    # future to avoid possible duplication of keys.
+    # TODO: Drop this in a future release
+    if '.' in key:
+        msg = ("a '.' was detected in the dict key: '{}' while formatting a dict passed to "
+               "ActionEvent.set_result(). Keys using the dotted notation will be rejected in a "
+               "later release.")
+        warnings.warn(msg.format(key), DeprecationWarning)
+
+    # TODO: Drop the period from this regex when the above DeprecationWarning is removed
+    key_pattern = re.compile(r'^[a-z0-9](([a-z0-9-.]+)?[a-z0-9])?$')
```

Comment:
I know Python's `re` module caches the first N items, so this probably isn't an issue in practice, but it's still good practice to call `re.compile` once at import time and store the result in a constant/variable. For example, see `METRIC_KEY_REGEX` elsewhere in this file.

Reply:
Will address in next commit, thanks!